### DISCUSSION TITLE
help with Non-Geospatial TileLayer Moon example on ObservableHQ

### DISCUSSION THREAD

**Discussion by cornhundred** [visgl/deck.gl#5346] [Q&A] [answered]
Hi, I'm trying to get the Non-Geospatial TileLayer Moon image example (https://deck.gl/examples/tile-layer-non-geospatial/, code: https://github.com/visgl/deck.gl/blob/8.3-release/examples/website/image-tile/app.js) working on ObservableHQ - notebook link [here](https://observablehq.com/@cornhundred/deck-gl-moon-tillayer-example). However, I am not experienced with React and am not sure how I can define [`setDimensions`](https://github.com/visgl/deck.gl/blob/8.3-release/examples/website/image-tile/app.js#L37) outside of React. Similarly, I am not sure where `onTilesLoad` is being defined and how this could be transferred to an Observable cell.

I was able to get the Geospatial TileLayer [example](https://deck.gl/docs/api-reference/geo-layers/tile-layer) working on Observable [here](https://observablehq.com/@cornhundred/geojson-tile-layer).

Any help would be appreciated :)

**Comment by cornhundred** [comment, marked as answer]
I was able to get the example working on Observable by separating the functions into separate cells - I still have to look over the documentation to make sure I understand everything that is going on but the example can be found here: https://observablehq.com/@cornhundred/deck-gl-moon-tillayer-example

**Comment by ilan-gold** [comment]
@cornhundred Very cool! I wonder if [Viv](https://github.com/hms-dbmi/viv) could be used in an observable notebook....@manzt check this out! I don't think I've ever seen you use Viv in a observable notebook before (although my memory is pretty terrible) but it would be cool!

**Reply by cornhundred** [reply]
Thanks @ilan-gold! It would be great to have Viv available on ObservableHQ, I'm happy to try and help out if you all want. Right now I'm working on swapping some biological image data for the moon image as a proof of concept. The performance of the moon example is great, I'm guessing in part because it is using JPEG compression. 

Observable is great for stringing different visuals together - I've been experimenting with interactions between spatial data (using deck.gl) and our interactive heatmap Clustergrammer-GL ([CitiBike data](https://observablehq.com/@cornhundred/citibike-clustergrammer-gl-and-deck-gl?collection=@cornhundred/clustergrammer-gl), [Visium spatial transcriptomics](https://observablehq.com/@cornhundred/clustergrammer-gl-visium-example?collection=@cornhundred/clustergrammer-gl), [NBA shot chart](https://observablehq.com/@cornhundred/nba-shot-chart-clustergrammer-gl?collection=@cornhundred/clustergrammer-gl)) and it would be great to have Viv available to work with.

Looking forward to checking out @manzt Observable notebooks too https://observablehq.com/@manzt/spatial-gene-expression-2m-molecules

**Comment by ilan-gold** [comment]
Very cool! I love the NBA shot chart (huge fan!).  I don't see why Viv's [deck.gl layers](https://github.com/hms-dbmi/viv/tree/master/src/layers) would not work in Observable, but first I need to get my chops up in Observable to find out....

**Reply by cornhundred** [reply]
Cool, I'll check out the Viv deck.gl layer :D

**Reply by manzt** [reply]
I don't think that there is anything fundamental that makes viv incompatible with observable, but the way it is currently packaged doesn't lend well to importing in the browser from a CDN.

Viv doesn't bundle any of its dependencies, so it will rely on some of observable's magic to resolve those dependencies. Some of its deps (geotiff.js) require a bit of webpack magic to get to run. I haven't tried in a while, but I remember hitting a wall early on because of this.

**Reply by ilan-gold** [reply]
geotiff.js does have a "browser" bundle that I have seen people report as functioning well but I personally have never tried using it....

**Reply by ilan-gold** [reply]
Although we are on an unbuilt fork of it for a few reasons so I'm not sure how that will work out....

**Reply by manzt** [reply]
Not debating whether or not it works well, but Observable would need to be able to resolve and "import" it somehow. If my recollection holds, I believe the browser bundle is UMD and the main export is commonjs (both generated via parcel). I don't know the details of observable's dependency resolver, but imports aren't really a thing from UMD (usually just assign a browser global) and the commonjs version usually has node-isms that set everything on fire.

Have a pure ESM export (without imports from node builtins) and then we're talking! 🔥🔥🔥

**Reply by manzt** [reply]
> Although we are on an unbuilt fork of it for a few reasons so I'm not sure how that will work out....

Yes. Observables resolver runs in the browser, so no way for it to build. It just connects static assets.

**Reply by cornhundred** [reply]
@ilan-gold @manzt is viv or it's custom layers available on npm? If so we could try this notebook to help import them https://observablehq.com/@observablehq/how-to-require-stubborn-modules

**Reply by cornhundred** [reply]
We've been experimenting with importing Observable notebook cells into Jupyter (using observable-jupyter). This allows us to use deck.gl in Jupyter when custom widgets are not supported (e.g. pydeck). Here's a scatterplot example on [Colab](https://colab.research.google.com/drive/1lXDh_AFgx3m9Qp307nz--1UwoBDkfbtH?usp=sharing). If viv works on Observable maybe something similar can be done on Colab.

**Reply by ilan-gold** [reply]
@cornhundred I think Trevor laid it out but some of our deps are not bundled as ESM modules (or at all) for browsers.  I think this is mainly isolated to geotiff.js since we are on a fork of it which lets us
1. Use a functioning LZW decoder (the one with the geotiff.js library crashes very easily)
2. Use AbortSignals to cancel ongoing requests

geotiff.js (the actual library, not my fork) has a UMD module so if you can get that working, and don't mind these features being missing, you should be good.  You can also pull my fork, build it, and push it to a CDN.  Let me know how it goes!

**Reply by cornhundred** [reply]
Thanks for the info @ilan-gold I'll let you know if I make progress.

**Reply by manzt** [reply]
If you run into a particularly difficult dependency, we might be able to look into splitting out npm distribution with separate entrypoints:

```bash
index.js    # top level export, includes all submodules
layers.js   # deck.gl primary dependency
loaders.js  # zarr.js & geotiff.js are dependencies
```

but my guess is that the loaders are also needed, not just the layers.

One thing to note about usage of Viv in Colab via observable. In the example you've shared, all data is loaded into memory and then converted to a list of dictionaries. Viv's layers extend deck.gl's image or `TileLayer` layers and load data tiles dynamically, in most cases using the `fetch` API and HTTP requests. To integrate `vizarr` in Jupyter Notebooks, I used the `imjoy-rpc` to let me dynamically load chunks from a python backend: https://colab.research.google.com/github/hms-dbmi/vizarr/blob/master/example/mandelbrot.ipynb

**Reply by cornhundred** [reply]
@manzt that's a very cool example! Someone on the ObservableHQ discourse pointed me to a method to send more data to the Observable notebook cell using [flask/ngrok](https://talk.observablehq.com/t/running-google-colab-from-observable-or-vice-versa/4868/3) and I made a proof of concept where I'm able to dynamically load columns from a dataset using a dropdown from Observable which makes a GET request to the Flask backend (which is public via ngrok) [Colab](https://colab.research.google.com/drive/18c6DkTLrbLtVDLnpvAaJUGw2uUdJJCyB?usp=sharing) - but this is a little hacky. It would be interesting to see if there's a more straightforward way to have front/backend communication without using Jupyter widgets (because of their lack of support for custom widgets on Google Colab).

**Reply by cornhundred** [reply]
@manzt have you all tried using viv as a layer in PyDeck, which currently does not support [tilelayers](https://github.com/visgl/deck.gl/discussions/5304#discussioncomment-265498)?

**Reply by ilan-gold** [reply]
@cornhundred The lack of support [was mentioned to me over a year ago](https://github.com/visgl/deck.gl/pull/4328#pullrequestreview-365849290) and I haven't looked into it at all since (nor did I look into it then).  There is interest in supporting that though, clearly!!

That being said, the `ImageLayer` (i.e not multiscale) or the incoming [3d volume raycasting layer](https://github.com/hms-dbmi/viv/pull/415) would both probably work.  I'm not sure how our loaders would play into this declarative style since they are JavaScript classes, but we export [the underlying rendering layer as well](http://viv.gehlenborglab.org/#xrlayer) (and will do so for 3d too). You really just need to pass in a typed array (like `Uint16Array`) to get those working, so the barrier to entry should be fairly low.

Hope this helps!

**Reply by cornhundred** [reply]
Thanks @ilan-gold thanks for the info! I'll try to figure out how these things are working :)
